### PR TITLE
C51-133: New milestones block

### PR DESCRIPTION
**Pull request by AkA84**
This PR adds the "milestones due" block in the dashboard, which displays activities based on these criteria:
* with grouping = to "milestones"
* with an incomplete status
* with an `activity_date_time` between the current date time and the end of the week/month (based on what period range it was selected

**With data**
milestones-data

**With no data**
milestones-no-data

The main bulk of the pr was simply about setting up the data necessary for the `query-panel` directive to do its job, and to refactor the code

### CODE REVIEW

File: ang/civicase/DashboardTab.js [hunk 1]
```diff
@@ -47,39 +75,79 @@
...
+          return;
+        }
+
+        $rootScope.$broadcast(
+          'civicaseActivityFeed.query',
+          $scope.newMilestonesPanel.custom.involvementFilter,
+          $scope.newMilestonesPanel.query.params,
+          true
+        );
+      }, true);
```
Comment (reneolivo) [inline ang/civicase/DashboardTab.js]: This is exactly why wrapper directives would be good.

File: ang/civicase/DashboardTab.js [hunk 2]
```diff
@@ -47,39 +75,79 @@
...
+     * Sets the range of the date of the entity (Case / Activity)
      *
+     * @param {String} property the property where the information about the date is stored
+     * @param {String} format the date format
+     * @param {Boolean} useNowAsStart whether the starting point should be the current datetime
      * @param {String} selectedRange the currently selected period range
      * @param {Object} queryParams
      */
-    function casesRangeHandler (selectedRange, queryParams) {
+    function rangeHandler (property, format, useNowAsStart, selectedRange, queryParams) {
```
Comment (reneolivo) [inline ang/civicase/DashboardTab.js]: Can we simplify the parameters in this function? I think using the curry function makes this more complicated and hides the fact where all these parameters come from. I'd consider returning a plain old handler function:

```js
function getHandlerForRangeChange(fieldName, options) {
  // options contain the format and useNowAsStart

  return function rangeChangeHandler (selectedRange, queryParams) {
    // ...
  }
}
```

File: ang/test/civicase/DashboardTab.spec.js [hunk 3]
```diff
@@ -208,12 +180,282 @@
...
+     *
+     * @param {String} range
+     * @param {String} format
+     * @return {Array}
+     */
+    function getStartEndOfRange (range, format, useNowAsStart) {
+      var now = moment();
+
+      return [
+        (useNowAsStart ? now : now.startOf(range)).format(format),
```
Comment (reneolivo) [inline ang/test/civicase/DashboardTab.spec.js]: can we refactor this bit here? I rather have this in a different line than mix it with all these brackets, parenthesis, etc.

Maybe:

File: ang/civicase/DashboardTab.html [hunk 4]
```diff
@@ -204,9 +204,35 @@
...
+            </div>
+          </panel-query-empty>
+        </civicase-panel-query>
+        <!-- New Milestones -->
+        <civicase-panel-query query="newMilestonesPanel.query" handlers="newMilestonesPanel.handlers" custom-data="newMilestonesPanel.custom">
+          <panel-query-title>
+            You have {{loading.full ? '-' : total}} milestones due
+          </panel-query-title>
+          <panel-query-actions>
+            <div ng-if="customData.involvementFilter" civicase-activity-filters-contact="customData.involvementFilter" class="panel-heading-control"></div>
```
[the remark this answers is not in the document]
Comment (AkA84) [reply ang/civicase/DashboardTab.html]: no, unfortunately given that the slots are being compiled on the directive's scope and placed inside the DOM via the link function, they need to be  present in the DOM right from the start

File: ang/civicase/DashboardTab.js [hunk 5]
```diff
@@ -11,25 +11,53 @@
...
     var CASES_QUERY_PARAMS_DEFAULTS = {
       'status_id.grouping': 'Opened',
       'options': { 'sort': 'start_date DESC' }
     };
+    var MILESTONES_QUERY_PARAMS_DEFAULTS = {
+      'contact_id': 'user_contact_id',
+      'is_current_revision': 1,
+      'is_deleted': 0,
+      'is_test': 0,
+      'activity_type_id.grouping': { 'LIKE': '%milestone%' },
```
[the remark this answers is not in the document]
Comment (AkA84) [reply ang/civicase/DashboardTab.js]: just following exactly what it is done in the "Activitied Feed" tab, no questions asked

**Pull request merged**